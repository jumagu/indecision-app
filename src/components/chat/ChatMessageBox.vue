<script setup lang="ts">
import { ref } from 'vue';

import SendIcon from '../icons/SendIcon.vue';

// Events definition
const emit = defineEmits<{ sendMessage: [text: string] }>();

const message = ref('');

const sendMessage = () => {
  if (!message.value.trim().length) return;

  emit('sendMessage', message.value);

  message.value = '';
};
</script>

<template>
  <div class="bg-white p-4 flex items-center">
    <input
      type="text"
      v-model="message"
      placeholder="Type your message..."
      class="flex-1 border rounded-full px-4 py-2 focus:outline-none"
      @keypress.enter="sendMessage"
    />
    <button
      class="bg-blue-500 text-white rounded-full p-2 ml-2 hover:bg-blue-600 focus:outline-none"
      @click="sendMessage"
    >
      <SendIcon />
    </button>
  </div>
</template>
