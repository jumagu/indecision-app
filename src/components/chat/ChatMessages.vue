<script setup lang="ts">
import { ref, watch } from 'vue';

import ThreeDotsIcon from '../icons/ThreeDotsIcon.vue';
import ChatMessageBubble from './ChatMessageBubble.vue';
import type { IChatMessage } from '@/interfaces/chat-message.interface';

interface IChatMessagesProps {
  isLoading?: boolean;
  messages: IChatMessage[];
}

const { messages } = defineProps<IChatMessagesProps>();

const messagesRef = ref(messages);
const msgContainerRef = ref<HTMLDivElement | null>(null);

watch(messagesRef.value, () => {
  setTimeout(() => {
    msgContainerRef.value?.scrollTo({
      top: msgContainerRef.value.scrollHeight,
      behavior: 'smooth',
    });
  }, 100);
});
</script>

<template>
  <div ref="msgContainerRef" class="flex-1 overflow-y-auto p-4">
    <div class="flex flex-col space-y-2">
      <ChatMessageBubble v-for="message in messages" :key="message.id" v-bind="message" />
      <div v-if="isLoading" class="flex justify-center">
        <ThreeDotsIcon width="64px" fill="#3b82f6" />
      </div>
    </div>
  </div>
</template>
