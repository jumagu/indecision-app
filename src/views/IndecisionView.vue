<script setup lang="ts">
import { useChat } from '@/composables/useChat';
import ChatMessages from '@/components/chat/ChatMessages.vue';
import ChatMessageBox from '@/components/chat/ChatMessageBox.vue';
import TrashCanIcon from '@/components/icons/TrashCanIcon.vue';

const { messages, isLoading, clearChat, sendMessageHandler } = useChat();
</script>

<template>
  <div class="bg-gray-100 h-screen flex flex-col max-w-lg mx-auto">
    <div class="bg-blue-500 p-4 text-white flex justify-between items-center">
      <div>
        <img
          class="inline-block h-9 w-9 rounded-full object-cover ring-2 ring-white"
          src="../../public/anonymous.jpg"
          alt="Anonymous"
        />
        <span class="ml-2">Anonymous</span>
      </div>

      <button
        class="bg-red-500 text-white rounded-full p-2 ml-2 hover:bg-red-600 focus:outline-none disabled:opacity-80 disabled:pointer-events-none"
        :disabled="!messages.length"
        @click="clearChat"
      >
        <TrashCanIcon />
      </button>
    </div>

    <!-- Messages Section -->
    <ChatMessages :messages="messages" :is-loading="isLoading" />

    <!-- Message Box -->
    <ChatMessageBox @send-message="sendMessageHandler" />
  </div>
</template>
