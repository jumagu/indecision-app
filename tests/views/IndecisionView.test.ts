import { mount } from '@vue/test-utils';

import IndecisionView from '@/views/IndecisionView.vue';
import ChatMessages from '@/components/chat/ChatMessages.vue';
import ChatMessageBox from '@/components/chat/ChatMessageBox.vue';

const chatMessagesMock = {
  template: '<div data-testid="messages-mock">Mock of ChatMessages component</div>',
};

describe('Tests in <IndecisionView />', () => {
  test('Should render chat messages and messagebox correctly', () => {
    const wrapper = mount(IndecisionView);
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.findComponent(ChatMessages).exists()).toBe(true);
    expect(wrapper.findComponent(ChatMessageBox).exists()).toBe(true);
  });

  test('Should call sendMessageHandler when sending a message', async () => {
    const wrapper = mount(IndecisionView, {
      global: {
        stubs: {
          ChatMessages: chatMessagesMock,
        },
      },
    });

    // const sendMessageHandlerMock = vi.fn();

    const messageboxComponent = wrapper.findComponent(ChatMessageBox);

    // (messageboxComponent.vm.$.parent?.vnode.component as any).setupState.sendMessageHandler =
    //   sendMessageHandlerMock;

    messageboxComponent.vm.$emit('sendMessage', 'Hello World');

    await new Promise((resolve) => setTimeout(resolve, 150));

    // expect(sendMessageHandlerMock).toHaveBeenCalled();
    // expect(messageboxComponent.vm.onSendMessage).toHaveBeenCalled();
  });
});
