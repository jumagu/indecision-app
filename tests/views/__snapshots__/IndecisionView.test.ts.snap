// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests in <IndecisionView /> > Should render chat messages and messagebox correctly 1`] = `
"<div class="bg-gray-100 h-screen flex flex-col max-w-lg mx-auto">
  <div class="bg-blue-500 p-4 text-white flex justify-between items-center">
    <div><img class="inline-block h-9 w-9 rounded-full object-cover ring-2 ring-white" src="/public/anonymous.jpg" alt="Anonymous"><span class="ml-2">Anonymous</span></div><button class="bg-red-500 text-white rounded-full p-2 ml-2 hover:bg-red-600 focus:outline-none disabled:opacity-80 disabled:pointer-events-none" disabled=""><svg height="20" width="20" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512" fill="currentColor">
        <g>
          <path fill="currentColor" d="M88.594,464.731C90.958,491.486,113.368,512,140.234,512h231.523c26.858,0,49.276-20.514,51.641-47.269 l25.642-335.928H62.952L88.594,464.731z M420.847,154.93l-23.474,307.496c-1.182,13.37-12.195,23.448-25.616,23.448H140.234 c-13.42,0-24.434-10.078-25.591-23.132L91.145,154.93H420.847z"></path>
          <path fill="currentColor" d="M182.954,435.339c5.877-0.349,10.35-5.4,9.992-11.269l-10.137-202.234c-0.358-5.876-5.401-10.349-11.278-9.992 c-5.877,0.357-10.35,5.409-9.993,11.277l10.137,202.234C172.033,431.231,177.085,435.696,182.954,435.339z"></path>
          <path fill="currentColor" d="M256,435.364c5.885,0,10.656-4.763,10.656-10.648V222.474c0-5.885-4.771-10.648-10.656-10.648 c-5.885,0-10.657,4.763-10.657,10.648v202.242C245.344,430.601,250.115,435.364,256,435.364z"></path>
          <path fill="currentColor" d="M329.046,435.339c5.878,0.357,10.921-4.108,11.278-9.984l10.129-202.234c0.348-5.868-4.116-10.92-9.993-11.277 c-5.877-0.357-10.92,4.116-11.277,9.992L319.054,424.07C318.697,429.938,323.17,434.99,329.046,435.339z"></path>
          <path fill="currentColor" d="M439.115,64.517c0,0-34.078-5.664-43.34-8.479c-8.301-2.526-80.795-13.566-80.795-13.566l-2.722-19.297 C310.388,9.857,299.484,0,286.642,0h-30.651H225.34c-12.825,0-23.728,9.857-25.616,23.175l-2.721,19.297 c0,0-72.469,11.039-80.778,13.566c-9.261,2.815-43.357,8.479-43.357,8.479C62.544,67.365,55.332,77.172,55.332,88.38v21.926h200.66 h200.676V88.38C456.668,77.172,449.456,67.365,439.115,64.517z M276.318,38.824h-40.636c-3.606,0-6.532-2.925-6.532-6.532 s2.926-6.532,6.532-6.532h40.636c3.606,0,6.532,2.925,6.532,6.532S279.924,38.824,276.318,38.824z"></path>
        </g>
      </svg></button>
  </div><!-- Messages Section -->
  <div class="flex-1 overflow-y-auto p-4">
    <div class="flex flex-col space-y-2">
      <!--v-if-->
    </div>
  </div><!-- Message Box -->
  <div class="bg-white p-4 flex items-center"><input type="text" placeholder="Type your message..." class="flex-1 border rounded-full px-4 py-2 focus:outline-none"><button class="bg-blue-500 text-white rounded-full p-2 ml-2 hover:bg-blue-600 focus:outline-none disabled:opacity-80 disabled:pointer-events-none" disabled=""><svg width="20" height="20" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg" stroke="#fff">
        <path d="M11.5003 12H5.41872M5.24634 12.7972L4.24158 15.7986C3.69128 17.4424 3.41613 18.2643 3.61359 18.7704C3.78506 19.21 4.15335 19.5432 4.6078 19.6701C5.13111 19.8161 5.92151 19.4604 7.50231 18.7491L17.6367 14.1886C19.1797 13.4942 19.9512 13.1471 20.1896 12.6648C20.3968 12.2458 20.3968 11.7541 20.1896 11.3351C19.9512 10.8529 19.1797 10.5057 17.6367 9.81135L7.48483 5.24303C5.90879 4.53382 5.12078 4.17921 4.59799 4.32468C4.14397 4.45101 3.77572 4.78336 3.60365 5.22209C3.40551 5.72728 3.67772 6.54741 4.22215 8.18767L5.24829 11.2793C5.34179 11.561 5.38855 11.7019 5.407 11.8459C5.42338 11.9738 5.42321 12.1032 5.40651 12.231C5.38768 12.375 5.34057 12.5157 5.24634 12.7972Z" stroke="#fff" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"></path>
      </svg></button></div>
</div>"
`;
